docs/engines.md

```diff
@@ -0,0 +1,209 @@
+accept engine specific configuration as an `EngineConfig` parameter. This allows
+for clients to set configuration data needed for the engine. The assumption is
+that the engine itself does not initialize the graphics API. The API should
+have been provided and all needed information provided in the `EngineConfig`
+object.
+
+A `Delegate` object is also provided to the engine. The delegate object is used
+when the engine needs to talk back to the embedder. For instance, the delegates
+can tell the engine to `LogGraphicsCalls`. The engine then, should, call the
+`Log` method on the delegate for each internal graphics API method called.
+
+If the executing script specified specific features, instance or device
+extensions they will also be provided. The engine can use these as needed.
+
+
+#### `CreatePipeline`
+The `CreatePipeline` method is responsible for creating an engine specific
+version of the given `amber::Pipeline`. Each command which needs a pipeline
+will have the pipeline provided, so the engine must provide a way to go from
+the amber pipeline to the engine pipeline. There can also be multiple pipeline
+lines of a given type.
+
+All information on the pipeline will be provided including shader and buffer
+information.
+
+
+#### `DoClearColor`
+The `DoClearColor` command provides the colour that is to be used for a given
+pipeline when executing the `DoClear` command.
+
+
+#### `DoClearStencil`
+The `DoClearStencil` command provides the value that is to be used for clearing
+the stencil buffer for a given pipeline when `DoClear` is executed.
+
+
+#### `DoClearDepth`
+The `DoClearDepth` command provides the value that is to be used for clearing
+the depth buffer for a given pipeline when `DoClear` is executed.
+
+
+#### `DoClear`
+The `DoClear` command instructs the engine to clear the various colour and
+depth attachments for the given pipeline.
+
+
+#### `DoDrawRect`
+The `DoDrawRect` instructs the engine to draw the given pipeline in the box
+at (x,y) of size (width,height). The buffers must be read back into the backing
+`amber::Buffer` at the end of this method.
+
+
+### `DoDrawArrays`
+The `DoDrawArrays` instructs the engine to draw the given pipeline using
+the information in the attached vertex and index buffers.  The buffers must be
+read back into the backing `amber::Buffer` at the end of this method.
+
+
+#### `DoCompute`
+The `DoCompute` instructs the engine to execute the given compute pipeline with
+the provided (x,y,z) parameters.
+
+
+#### `DoEntryPoint`
+The `DoEntryPoint` command instructs the engine to change the entry point in
+the given pipeline for the give shader.
+
+
+#### `DoPatchParameterVertices`
+The `DoPatchParameterVertices` tells the engine to do a thing with some
+values....I don't really know what it means, heh.
+
+
+#### `DoBuffer`
+The `DoBuffer` command tells the engine that for a given pipeline the given
+data should be updated into the given buffer.
+
+
+### `SetEngineData`
```

Comment:
This one is at a higher level of header. (3 rather than 4)

Reply:
done